src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -446,22 +483,89 @@ trait OptimisticTransactionImpl extends TransactionalWrite {
           case _ =>
         }
       }
+
       // Fail if the metadata is different than what the txn read.
       if (metadataUpdates.nonEmpty) {
         throw new MetadataChangedException(commitInfo)
       }
-      // Fail if the data is different than what the txn read.
-      if (dependsOnFiles && fileActions.nonEmpty) {
-        throw new ConcurrentWriteException(commitInfo)
+
+      // Fail if new files have been added that the txn should have read.
+      val addedFilesToCheckForConflicts = commitIsolationLevel match {
+        case Serializable => changedDataAddedFiles ++ blindAppendAddedFiles
+        case WriteSerializable => changedDataAddedFiles // don't conflict with blind appends
+        case SnapshotIsolation => Seq.empty
       }
+      val predicatesMatchingAddedFiles = ExpressionSet(readPredicates).iterator.flatMap { p =>
+        val conflictingFile = DeltaLog.filterFileList(
+          metadata.partitionColumns,
+          addedFilesToCheckForConflicts.toDF(), p :: Nil).as[AddFile].take(1)
+
+        conflictingFile.headOption.map(f => getPrettyPartitionMessage(f.partitionValues))
+      }.take(1).toArray
+
+      if (predicatesMatchingAddedFiles.nonEmpty) {
+        val isWriteSerializable = commitIsolationLevel match {
+          case WriteSerializable => true
+          case _ => false
+        }
```

Comment:
nit: should we just do commitIsolationLevel == WriteSerializable below? this is pretty verbose